sql/api/src/main/scala/org/apache/spark/sql/api/Dataset.scala

```diff
@@ -2595,7 +2595,7 @@ abstract class Dataset[T] extends Serializable {
    * @group typedrel
    * @since 1.6.0
    */
-  def transform[U](t: Dataset[T] => Dataset[U]): Dataset[U] = t(this.asInstanceOf[Dataset[T]])
+  def transform[U](t: DS[T] => DS[U]): DS[U] = t(this.asInstanceOf[DS[T]])
```

Comment:
I'd sort of prefer this to return `Dataset[U]` and use covariant overloads in the implementations. See join/union for example. It is a tad more consistent.